replicators/src/mysql_connector/snapshot.rs

```diff
@@ -665,6 +680,41 @@ impl MySqlReplicator {
 
         Ok(())
     }
+
+    /// This function drops all known tables from Noria. This is used when we require a full
+    /// re-snapshot of the database.
+    ///
+    /// # Arguments
+    /// * `noria`: The target ReadySet deployment
+    async fn drop_all_tables(
+        &mut self,
+        noria: &mut readyset_client::ReadySetHandle,
+    ) -> ReadySetResult<()> {
+        let replication_offsets = noria.replication_offsets().await?;
+        let mut tx = self.pool.start_transaction(tx_opts()).await?;
+        let all_tables = get_table_list(&mut tx, TableKind::BaseTable).await?;
+        let mut changes = Vec::new();
+        for (db, table) in all_tables.iter() {
+            let table = Relation {
+                schema: Some(db.clone().into()),
+                name: table.clone().into(),
+            };
+            if replication_offsets.has_table(&table) {
+                changes.push(Change::Drop {
+                    name: table,
+                    if_exists: true,
+                });
+            }
+        }
+        tx.rollback().await?;
+        noria
+            .extend_recipe_no_leader_ready(ChangeList::from_changes(
+                changes,
+                Dialect::DEFAULT_MYSQL,
+            ))
+            .await?;
```

Comment:
I'd like someone else (@ethan-readyset or @lukoktonos perhaps) to take a look at this bit in particular to double-check my intuition that there aren't likely to be any race conditions here - specifically I don't *think* we get into any trouble here if the transaction we've started is after some `DROP TABLE` or `CREATE TABLE` DDL event we've yet to replicate, but I'm not sure.

Reply:
I am not very familiar with our MySQL replication code, but here's a quick overview of how I understand it based on a code dive:
* We attempt to do a `LOCK INSTANCE FOR BACKUP`, but this is not supported by RDS, so it ends up being a noop for most users
* We start a readonly transaction with `CONSISTENT SNAPSHOT` and isolation level "repeatable read"
* We get a list of all the tables
* We get `CREATE TABLE` statements for those tables via `SHOW CREATE`
* We repeat the above two steps for views
* For each table, we:
  * Acquire a read lock via `LOCK READ`
  * Read the replication offset
  * Snapshot the table into ReadySet in the context of a newly-created transaction for each table
  * Set the replication offset
  * Compact the table

I'm not super sure why we need to acquire per-table read locks and create a new per-table transaction if we've already created that top-level transaction, within which we should have a consistent view of all the data in the database at the time we started the transaction.

Regardless, for good hygiene, the new `drop_all_tables` method should probably use the same transaction as the one created by `load_recipe_with_meta_lock` -- that way, we can be sure that the state of the world as seen by those two methods is the same.

Reply:
Hi @ethan-readyset 
> I'm not super sure why we need to acquire per-table read locks and create a new per-table transaction if we've already created that top-level transaction, within which we should have a consistent view of all the data in the database at the time we started the transaction.

The upper transaction tries to acquire LIFB, which prevents new tables from being created or modifying the existing ones. As you mentioned  RDS does not allow this Metadata Lock (MDL). What in fact RS does:
* Start the new transaction
* Get the list of the tables
* Inside the transaction, execute SELECT 1 FROM (table list) LIMIT 0 - This will add a MDL on those tables preventing their structure from changing. Still, this does not cover the scenario of new tables arriving when LIFB is not supported. This transaction is held open for the duration of the snapshot.

Now for each discovered table RS snapshot them in parallel (I can see it allows to up 10 parallel threads). RS has one trx that is responsible for data synchronization (T0) and one that is responsible for dumping data (T1):
* T0: Acquire a read lock via LOCK READ (this is done just for a brief moment )
* T0: Read the replication offset
* T1: Start a readonly transaction with CONSISTENT SNAPSHOT and isolation level "repeatable read" (at this moment is guaranteed that T0 replication offsets and snapshot of data taken by T1 are in sync as the table has READ LOCK
* T0: Unlock table
* T1: SELECT * FROM table (start to snapshot the table)
For the next table, T0 will perform the same steps and we will have a T2, T3 responsible for the snapshot of the individual tables.

Reply:
> Regardless, for good hygiene, the new drop_all_tables method should probably use the same transaction as the one created by load_recipe_with_meta_lock -- that way, we can be sure that the state of the world as seen by those two methods is the same.

Make sense to synchronize them. What about changing `drop_all_tables` to take a table list as parameter and invoke it inside `load_recipe_with_meta_lock` right after we get the table list [here](https://github.com/readysettech/readyset/blob/beta-2023-08-23/replicators/src/mysql_connector/snapshot.rs#L166) ?

Reply:
@altmannmarcelo re: your first comment, I think my primary confusion is that I don't think we need to prevent changes from happening on the upstream database -- we just need to make sure that when we read data from the database during the snapshotting process, all of the queries we issue are reading the database from a consistent point in time. This is how we handle snapshotting a Postgres database without using any locks, so I'm not sure the history as to why we need to acquire locks when snapshotting a MySQL database. Isn't it enough just to have one transaction with `CONSISTENT SNAPSHOT` and isolation level "repeatable read" to ensure that we have a consistent view of the database at a particular point in time? If any writes occur while we're snapshotting, we'd replicate those writes after we finish snapshotting.

This conversation is probably out of scope for this PR, so we don't need to dig much further into this here -- I was just mostly thinking out loud 😄 

>Make sense to synchronize them. What about changing drop_all_tables to take a table list as parameter and invoke it inside load_recipe_with_meta_lock right after we get the table list [here](https://github.com/readysettech/readyset/blob/beta-2023-08-23/replicators/src/mysql_connector/snapshot.rs#L166) ?

That's fine with me!

Reply:
@ethan-readyset I don't know the reason either :) . MySQL does not provide synchronization between START TRANSACTION WITH CONSISTENT SNAPSHOT and SHOW MASTER STATUS (8.0 actually has log_status table, but that is different topic ). I guess that is the reason RS shortly acquired READ LOCK. Other tools like mydumper do the same.

Reply:
>Make sense to synchronize them. What about changing drop_all_tables to take a table list as parameter and invoke it inside load_recipe_with_meta_lock right after we get the table list [here](https://github.com/readysettech/readyset/blob/beta-2023-08-23/replicators/src/mysql_connector/snapshot.rs#L166) ?

@altmannmarcelo Hmm was this change implemented? It still looks like we are invoking `drop_all_tables` outside of the context of the transaction in `load_recipe_with_meta_lock` and without an explicit table list

Reply:
@ethan-readyset  - This has evolved after [this conversation](https://github.com/readysettech/readyset/pull/530/files#diff-9d28be097addf1353c0f659cc635eef1ac7bb64b3afe8392e68d4bd1e42084c2) now we fetch the list of tables to drop from noria rather than the database. This seems the right thing to do, as we might have missed some DDLs that could potentially have dropped a table in some position ahead of the one triggering full_snapshot. In that case, we will have a dangling table in Noria.
Function drop_all_tables is being executed right after LOCK INSTANCE FOR BACKUP. However, this is not relevant as the binlog dump thread is stopped, and we won't receive any new event, meaning there is no room for race conditions here.
If you think we should execute the drop_all_tables somewhere else, I can move it to the desired place.

Reply:
FYI: I did a further FP to adjust the commit message as it was still referring to drop_single_table.

Reply:
Ah, I see now! Thanks for the clarification!